### PR TITLE
Added methods, classes to the solution:

### PR DESCRIPTION
**Pull request by nikolasdymchenko7**
- added class RegistrationValidatorException
- added implementation to the method RegistrationServiceImpl
- added class with JUnit tests for the RegistrationServiceImpl class

### CODE REVIEW

File: src/main/java/core/basesyntax/service/RegistrationService.java [hunk 1]
```diff
@@ -1,7 +1,8 @@
 package core.basesyntax.service;
 
 import core.basesyntax.model.User;
+import core.basesyntax.validators.RegistrationValidatorException;
 
 public interface RegistrationService {
-    User register(User user);
+    User register(User user) throws RegistrationValidatorException;
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/service/RegistrationService.java:7]: revert this change

File: src/main/java/core/basesyntax/validators/RegistrationValidatorException.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package core.basesyntax.validators;
+
+public class RegistrationValidatorException extends Exception {
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/validators/RegistrationValidatorException.java:3]: image

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 3]
```diff
@@ -3,12 +3,40 @@
...
     @Override
-    public User register(User user) {
-        return null;
+    public User register(User user) throws RegistrationValidatorException {
+        if (user == null) {
+            throw new RegistrationValidatorException("User shouldn't be null");
+        }
+        if (user.getLogin() == null || user.getLogin().length() < 5) {
+            throw new RegistrationValidatorException(
+                    "Login is null or length is less than 5 symbols"
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:20]: why 5 symbols? Where is it in the task?

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 4]
```diff
@@ -3,12 +3,40 @@
...
+            );
+        }
+        if (user.getLogin().startsWith(" ")) {
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:23]: Where is it in the task?

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 5]
```diff
@@ -3,12 +3,40 @@
...
+        if (user.getPassword() == null
+                || user.getPassword().length() < MIN_PASSWORD_CHARACTERS) {
+            throw new RegistrationValidatorException(
+                    "Your password is null or less than 5 characters"
+            );
+        }
+        if (user.getAge() == null || user.getAge() < MIN_AGE) {
+            throw new RegistrationValidatorException("Your age is null or less than 18 years old");
+        }
+        if (storageDao.get(user.getLogin()) == user) {
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:37]: think about this check and fix it

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 6]
```diff
@@ -3,12 +3,40 @@
...
+            throw new RegistrationValidatorException("User existing in the storage");
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:38]: make msg informative

File: src/main/java/core/basesyntax/validators/RegistrationValidatorException.java [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package core.basesyntax.validators;
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/validators/RegistrationValidatorException.java:1]:
```suggestion
package core.basesyntax.exception ;
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+    @Test
+    void register_nullUser_notOk() {
+        assertThrows(RegistrationValidatorException.class, () -> {
+            registrationService.register(null);
+        });
+    }
+
+    @Test
+    void register_nullLogin_notOk() {
+        testUser.setLogin(null);
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:35]: testUser has all fields null - look at init() method
set all correct fiels in init() and than set one incorrect in test methods.

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+    private static final int AGE_LESS_MIN = 17;
+    private static final int AGE_LESS_ZERO = -10;
+    private RegistrationServiceImpl registrationService;
+    private StorageDaoImpl storageDao;
+    private User testUser;
+
+    @BeforeEach
+    void init() {
+        registrationService = new RegistrationServiceImpl();
+        storageDao = new StorageDaoImpl();
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:22]: do we need new RegistrationServiceImpl and new StorageDaoImpl BeforeEach?

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 10]
```diff
@@ -2,13 +2,40 @@
...
+            throw new RegistrationValidatorException(
+                    "Your password is null or less than " + MIN_PASSWORD_CHARACTERS + " characters"
+            );
+        }
+        if (user.getAge() == null || user.getAge() < MIN_AGE) {
+            throw new RegistrationValidatorException("Your age is null or less than "
+                    + MIN_AGE + " years old");
+        }
+        if (storageDao.get(user.getLogin()) != null
+                && storageDao.get(user.getLogin()).getLogin().equals(user.getLogin())) {
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:34]:
```suggestion
        if (storageDao.get(user.getLogin()) != null) {
```

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 11]
```diff
@@ -2,13 +2,40 @@
...
+            throw new RegistrationValidatorException(
+                    "You can't add user because user exists in the storage!"
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:36]:
```suggestion
                    "Login already exists in the storage: " + user.getLogin()
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+    @BeforeAll
+    static void beforeAll() {
+        registrationService = new RegistrationServiceImpl();
+        storageDao = new StorageDaoImpl();
+    }
+
+    @BeforeEach
+    void init() {
+        testUser = new User();
+        testUser.setId(Long.valueOf("1"));
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:29]:
```suggestion
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+    void register_existingUserInStorage_notOk() {
+        storageDao.add(testUser);
+
+        assertThrows(RegistrationValidatorException.class, () -> {
+            registrationService.register(testUser);
+        });
+    }
+
+    @Test
+    void register_addNewUserInStorage_ok() throws RegistrationValidatorException {
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:61]:
```suggestion
    void register_addNewUserInStorage_ok() {
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+        testUser.setId(Long.valueOf(2));
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:62]:
```suggestion
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+    void register_ageLessZero_notOk() {
+        testUser.setAge(AGE_LESS_ZERO);
+        assertThrows(RegistrationValidatorException.class, () -> {
+            registrationService.register(testUser);
+        });
+    }
+
+    @Test
+    void register_age18_ok() throws RegistrationValidatorException {
+        testUser.setId(Long.valueOf(4));
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:98]:
```suggestion
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+        testUser.setPassword("12345");
+
+        assertThrows(RegistrationValidatorException.class, () -> {
+            registrationService.register(testUser);
+        });
+    }
+
+    @Test
+    void register_passwordWithSixCharacters_ok() throws RegistrationValidatorException {
+        testUser.setId(Long.valueOf(3));
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:127]:
```suggestion
```

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+    void register_nullAge_notOk() {
+        testUser.setAge(null);
+
+        assertThrows(RegistrationValidatorException.class, () -> {
+            registrationService.register(testUser);
+        });
+    }
+
+    @Test
+    void register_ageLessThan18_notOk() {
```
Comment (aaananas) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:78]: dont put concrete values into method names, those required values can be changed in the future

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+    void register_passwordWithFiveCharacters_notOk() {
+        testUser.setPassword("12345");
+
+        assertThrows(RegistrationValidatorException.class, () -> {
+            registrationService.register(testUser);
+        });
+    }
+
+    @Test
+    void register_passwordWithSixCharacters_ok() {
```
Comment (aaananas) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:123]: concrete value

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+    @Test
+    void register_ageLessZero_notOk() {
+        testUser.setAge(AGE_LESS_ZERO);
+        assertThrows(RegistrationValidatorException.class, () -> {
+            registrationService.register(testUser);
+        });
+    }
+
+    @Test
+    void register_age18_ok() {
```
Comment (aaananas) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:95]: concrete value

**Review by aaananas — CHANGES_REQUESTED** [review]

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+    @Test
+    void register_nullUser_notOk() {
+        assertThrows(RegistrationValidatorException.class, () -> {
+            registrationService.register(EMPTY_USER);
+        });
+    }
+
+    @Test
+    void register_nullLogin_notOk() {
+        testUser.setLogin(NULL_STRING);
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:54]:
```suggestion
        testUser.setLogin(null);
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+    private static final String DEFAULT_LOGIN = "nicolas";
+    private static final String LOGIN_FOR_NEW_USER = "bob";
+    private static final String LOGIN_FOR_CORRECT_AGE = "ben";
+    private static final String LOGIN_WITH_CORRECT_PASSWORD = "john";
+    private static final String DEFAULT_PASSWORD = "123456";
+    private static final String PASSWORD_LESS_MIN_BORDER = "12345";
+    private static final int DEFAULT_AGE = 20;
+    private static final User EMPTY_USER = null;
+    private static final String NULL_STRING = null;
+    private static final Integer NULL_AGE = null;
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:25]:
```suggestion
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+        testUser.setLogin(LOGIN_FOR_NEW_USER);
+        User expectedUser = testUser;
+        User actualUser = registrationService.register(testUser);
+
+        assertEquals(expectedUser, actualUser, "User has not been added to the storage");
+    }
+
+    @Test
+    void register_nullAge_notOk() {
+        testUser.setAge(NULL_AGE);
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:81]:
```suggestion
        testUser.setAge(null);
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java (hunk 17 above)
Comment (Rommelua) [reply src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:89]: not fixed

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+    void register_ageLessZero_notOk() {
+        testUser.setAge(AGE_LESS_ZERO);
+        assertThrows(RegistrationValidatorException.class, () -> {
+            registrationService.register(testUser);
+        });
+    }
+
+    @Test
+    void register_age18_ok() {
+        testUser.setLogin(LOGIN_FOR_CORRECT_AGE);
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:107]: What is the reason to change login?
Just do not forget to clean your storage after each test.
```suggestion
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+        Integer expectedAge = testUser.getAge();
+        Integer actualAge = registrationService.register(testUser).getAge();
+
+        assertEquals(expectedAge, actualAge, "Added user has incorrect age");
+    }
+
+    @Test
+    void register_nullPassword_notOk() {
+        testUser.setPassword(NULL_STRING);
+
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:118]:
```suggestion
        testUser.setPassword(null);
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 25]
```diff
@@ -0,0 +1,143 @@
...
+        testUser.setPassword(PASSWORD_LESS_MIN_BORDER);
+
+        assertThrows(RegistrationValidatorException.class, () -> {
+            registrationService.register(testUser);
+        });
+    }
+
+    @Test
+    void register_passwordWithSixCharacters_ok() {
+        testUser.setLogin(LOGIN_WITH_CORRECT_PASSWORD);
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:135]:
```suggestion
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 26]
```diff
@@ -0,0 +1,143 @@
...
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class RegistrationServiceImplTest {
+    private static final int AGE_LESS_MIN = 17;
+    private static final int AGE_LESS_ZERO = -10;
+    private static final String DEFAULT_LOGIN = "nicolas";
+    private static final String LOGIN_FOR_NEW_USER = "bob";
+    private static final String LOGIN_FOR_CORRECT_AGE = "ben";
+    private static final String LOGIN_WITH_CORRECT_PASSWORD = "john";
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:19]:
```suggestion
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java (hunk 19 above)
Comment (Rommelua) [reply src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:106]: not fixed

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java (hunk 18 above)
Comment (Rommelua) [reply src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:134]: not fixed

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 27]
```diff
@@ -0,0 +1,141 @@
...
+    void init() {
+        testUser = new User();
+        testUser.setLogin(DEFAULT_LOGIN);
+        testUser.setPassword(DEFAULT_PASSWORD);
+        testUser.setAge(DEFAULT_AGE);
+    }
+
+    @AfterEach
+    void cleanUpEach() {
+        Storage.people.remove(testUser);
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:43]:
```suggestion
        Storage.people.clear();
```

**Review by Rommelua — APPROVED** [review]
